### PR TITLE
[11.0][FIX] mrp: Consider stock move lines from products that have tracking

### PR DESCRIPTION
**Pull request by MiquelRForgeFlow**
The creation of stock move lines in the migration of `stock` and `mrp` modules was (and still is) a headache. I think it will never be perfect.

But I tried. This PR improves it a little bit more.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Can you explain the change and if this is linked to any issue?

**Comment by MiquelRForgeFlow** [comment]
When you have tracking By Lots (or By Unique Serial Number) like this:

Selection_072

Then, it happens the following:

v10:
Selection_071

v10 migrated v11 (before this PR):
Selection_069

v10 migrated v11 (with this PR):
Selection_070

File: addons/mrp/migrations/11.0.2.0/post-migration.py [hunk 1]
```diff
@@ -120,10 +120,20 @@ def create_stock_move_lines_from_stock_move_lots(env):
             sm.restrict_partner_id,
             sm.picking_id,
             sm.product_id,
-            0,
-            sm.product_uom,
-            0,
-            sm.product_uom_qty as qty_done,
+            CASE WHEN sm.is_done = TRUE THEN 0
```
Comment (pedrobaeza) [inline addons/mrp/migrations/11.0.2.0/post-migration.py]:
```suggestion
            CASE WHEN sml.lot_id IS NOT NULL THEN sml.quantity AND NOT sm.is_done THEN sml.quantity
            ELSE 0
```

File: addons/mrp/migrations/11.0.2.0/post-migration.py [hunk 2]
```diff
@@ -120,10 +120,20 @@ def create_stock_move_lines_from_stock_move_lots(env):
...
+                 WHEN sml.lot_id IS NOT NULL THEN sml.quantity
+            ELSE 0
+            END as product_qty,
+            CASE WHEN sml.lot_id IS NOT NULL THEN spl.product_uom_id
+            ELSE sm.product_uom
+            END as product_uom_id,
+            CASE WHEN sm.is_done = TRUE THEN 0
```
Comment (pedrobaeza) [inline addons/mrp/migrations/11.0.2.0/post-migration.py]: Do something similar as proposed above

File: addons/mrp/migrations/11.0.2.0/post-migration.py [hunk 3]
```diff
@@ -120,10 +120,20 @@ def create_stock_move_lines_from_stock_move_lots(env):
...
+                 WHEN sml.lot_id IS NOT NULL THEN sml.quantity
+            ELSE 0
+            END as product_uom_qty,
+            CASE WHEN sml.lot_id IS NOT NULL THEN sml.quantity_done
```
Comment (pedrobaeza) [inline addons/mrp/migrations/11.0.2.0/post-migration.py:134]: Do something similar as proposed above

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by pedrobaeza** [comment]
Is this for fixing #1952?

File: addons/mrp/migrations/11.0.2.0/post-migration.py (hunk 3 above)
Comment (MiquelRForgeFlow) [reply addons/mrp/migrations/11.0.2.0/post-migration.py:134]: but here cannot. Here there is only one WHEN hahaha

**Comment by MiquelRForgeFlow** [comment]
Maybe. I don't know if this fixes that or not. I just wanted to fix my own issue, nor the other ones. But it's possible that the one you mentioned is fixed with this one.

**Comment by pedrobaeza** [comment]
@OthmaneGhandi please try this fix for the problem you mentioned

**Comment by OthmaneGhandi** [comment]
@pedrobaeza & @mreficent  I will test it tomorrow and share with you my results

Thanks guys

**Comment by OthmaneGhandi** [comment]
@pedrobaeza & @mreficent, it solves the issue BUT it has some side effects:

**1/** I think we should keep 
**`WHERE sm.state NOT IN ('cancel', 'confirmed')`** 
instead of 
`WHERE sm.state NOT IN ('cancel')` 

Because with the last one it creates stock.move.line for the confirmed stock.moves (that are not done yet), it's like they considered the quantity was produced, which is not the case.

**2/** I had an issue (I don't know if it's specific only for me or it can happens to others) some stock.move.line was not related to the production record because in stock.move.lots the field production_id was empty (is there cases where produced products can have this field empty in stock.move.lot ?) that's why I tried to use: 

`sm.production_id,` 
instead of 
`sml.production_id` 

but it's not working too :-(

**Comment by OthmaneGhandi** [comment]
It works when you use : 

`from_ = """stock_move sm
        INNER JOIN mrp_production mp ON sm.production_id = mp.id
        LEFT JOIN stock_move_lots sml ON (sml.move_id = sm.id)
        LEFT JOIN stock_production_lot spl ON sml.lot_id = spl.id"""`

**instead of** 

`from_ = """stock_move sm
        INNER JOIN mrp_production mp ON sm.production_id = mp.id
        LEFT JOIN stock_move_lots sml ON (sml.move_id = sm.id
            AND sml.production_id = mp.id)
        LEFT JOIN stock_production_lot spl ON sml.lot_id = spl.id"""`

**Comment by MiquelRForgeFlow** [comment]
@OthmaneGhandi thanks for your suggestions. It helped me a lot. I think both you and I can be happy know with this PR :)

**Comment by OthmaneGhandi** [comment]
@mreficent you are very welcome, big thanks to you too for the great work.

just one remark I saw that you kept **`sml.production_id`** for me it worked with **`sm.production_id`** because some `stock.move.lots` the `production_id` empty : I don't know if it's a common case or is just happened to me

**Comment by MiquelRForgeFlow** [comment]
well, I can put `COALESCE(sml.production_id, sm.production_id)` if your prefer

**Comment by MiquelRForgeFlow** [comment]
@pedrobaeza ready to merge

**Review by pedrobaeza — APPROVED** [review]
FYI, there's no need to put `AS field_name`, as the assignment is done positional

**Comment by MiquelRForgeFlow** [comment]
I know. But I like to put it for readability. It you don't put it, sometimes you don't know what corresponds and you have to check the 'insert' part. But in this code, the 'insert' part is separated, several lines above, and it was a bit annoying to check.

**Comment by pedrobaeza** [comment]
OK, not too much problem any way, except longer lines that maybe don't fit in one line.

**Pull request merged**